lib/src/constants.dart

```diff
@@ -403,7 +403,7 @@ const spaceCharacters = ' \n\r\t\u000C';
 const int NEWLINE = 10;
 const int RETURN = 13;
 
-bool isWhitespace(String char) {
+bool isWhitespace(String? char) {
```

Comment:
I'd probably make this function not accept `null`.

Reply:
This is another one that saves a lot of noise.

```diff
diff --git a/lib/src/constants.dart b/lib/src/constants.dart
index 52c08d7..fd2f4d8 100644
--- a/lib/src/constants.dart
+++ b/lib/src/constants.dart
@@ -403,8 +403,7 @@ const spaceCharacters = ' \n\r\t\u000C';
 const int NEWLINE = 10;
 const int RETURN = 13;
 
-bool isWhitespace(String? char) {
-  if (char == null) return false;
+bool isWhitespace(String char) {
   return isWhitespaceCC(char.codeUnitAt(0));
 }
 
diff --git a/lib/src/tokenizer.dart b/lib/src/tokenizer.dart
index acd388f..5ba34a6 100644
--- a/lib/src/tokenizer.dart
+++ b/lib/src/tokenizer.dart
@@ -260,7 +260,7 @@ class HtmlTokenizer implements Iterator<Token> {
     String? output = '&';
 
     final charStack = [stream.char()];
-    if (isWhitespace(charStack[0]) ||
+    if (charStack[0] != null && isWhitespace(charStack[0]!) ||
         charStack[0] == '<' ||
         charStack[0] == '&' ||
         charStack[0] == eof ||
@@ -411,7 +411,7 @@ class HtmlTokenizer implements Iterator<Token> {
     } else if (data == eof) {
       // Tokenization ends.
       return false;
-    } else if (isWhitespace(data)) {
+    } else if (data != null && isWhitespace(data)) {
       // Directly after emitting a token you switch back to the "data
       // state". At that point spaceCharacters are important so they are
       // emitted separately.
@@ -445,7 +445,7 @@ class HtmlTokenizer implements Iterator<Token> {
     } else if (data == '\u0000') {
       _addToken(ParseErrorToken('invalid-codepoint'));
       _addToken(CharactersToken('\uFFFD'));
-    } else if (isWhitespace(data)) {
+    } else if (data != null && isWhitespace(data)) {
       // Directly after emitting a token you switch back to the "data
       // state". At that point spaceCharacters are important so they are
       // emitted separately.
@@ -567,7 +567,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool tagNameState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       state = beforeAttributeNameState;
     } else if (data == '>') {
       emitCurrentToken();
@@ -622,7 +622,7 @@ class HtmlTokenizer implements Iterator<Token> {
   bool rcdataEndTagNameState() {
     final appropriate = _tokenIsAppropriate();
     final data = stream.char();
-    if (isWhitespace(data) && appropriate) {
+    if (data != null && isWhitespace(data) && appropriate) {
       currentToken = EndTagToken('$_buffer');
       state = beforeAttributeNameState;
     } else if (data == '/' && appropriate) {
@@ -671,7 +671,7 @@ class HtmlTokenizer implements Iterator<Token> {
   bool rawtextEndTagNameState() {
     final appropriate = _tokenIsAppropriate();
     final data = stream.char();
-    if (isWhitespace(data) && appropriate) {
+    if (data != null && isWhitespace(data) && appropriate) {
       currentToken = EndTagToken('$_buffer');
       state = beforeAttributeNameState;
     } else if (data == '/' && appropriate) {
@@ -723,7 +723,7 @@ class HtmlTokenizer implements Iterator<Token> {
   bool scriptDataEndTagNameState() {
     final appropriate = _tokenIsAppropriate();
     final data = stream.char();
-    if (isWhitespace(data) && appropriate) {
+    if (data != null && isWhitespace(data) && appropriate) {
       currentToken = EndTagToken('$_buffer');
       state = beforeAttributeNameState;
     } else if (data == '/' && appropriate) {
@@ -863,7 +863,7 @@ class HtmlTokenizer implements Iterator<Token> {
   bool scriptDataEscapedEndTagNameState() {
     final appropriate = _tokenIsAppropriate();
     final data = stream.char();
-    if (isWhitespace(data) && appropriate) {
+    if (data != null && isWhitespace(data) && appropriate) {
       currentToken = EndTagToken('$_buffer');
       state = beforeAttributeNameState;
     } else if (data == '/' && appropriate) {
@@ -885,7 +885,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool scriptDataDoubleEscapeStartState() {
     final data = stream.char();
-    if (isWhitespace(data) || data == '/' || data == '>') {
+    if (data != null && isWhitespace(data) || data == '/' || data == '>') {
       _addToken(CharactersToken(data));
       if ('$_buffer'.toLowerCase() == 'script') {
         state = scriptDataDoubleEscapedState;
@@ -985,7 +985,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool scriptDataDoubleEscapeEndState() {
     final data = stream.char();
-    if (isWhitespace(data) || data == '/' || data == '>') {
+    if (data != null && isWhitespace(data) || data == '/' || data == '>') {
       _addToken(CharactersToken(data));
       if ('$_buffer'.toLowerCase() == 'script') {
         state = scriptDataEscapedState;
@@ -1004,7 +1004,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool beforeAttributeNameState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       stream.charsUntil(spaceCharacters, true);
     } else if (isLetter(data)) {
       _addAttribute(data);
@@ -1046,7 +1046,7 @@ class HtmlTokenizer implements Iterator<Token> {
       // without being checked and when the code below runs we error
       // because data is a dict not a list
       emitToken = true;
-    } else if (isWhitespace(data)) {
+    } else if (data != null && isWhitespace(data)) {
       state = afterAttributeNameState;
     } else if (data == '/') {
       state = selfClosingStartTagState;
@@ -1093,7 +1093,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool afterAttributeNameState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       stream.charsUntil(spaceCharacters, true);
     } else if (data == '=') {
       state = beforeAttributeValueState;
@@ -1124,7 +1124,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool beforeAttributeValueState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       stream.charsUntil(spaceCharacters, true);
     } else if (data == '"') {
       _markAttributeValueStart(0);
@@ -1207,7 +1207,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool attributeValueUnQuotedState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       _markAttributeValueEnd(-1);
       state = beforeAttributeNameState;
     } else if (data == '&') {
@@ -1235,7 +1235,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool afterAttributeValueState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       state = beforeAttributeNameState;
     } else if (data == '>') {
       emitCurrentToken();
@@ -1472,7 +1472,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool doctypeState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       state = beforeDoctypeNameState;
     } else if (data == eof) {
       _addToken(ParseErrorToken('expected-doctype-name-but-got-eof'));
@@ -1489,7 +1489,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool beforeDoctypeNameState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       return true;
     } else if (data == '>') {
       _addToken(ParseErrorToken('expected-doctype-name-but-got-right-bracket'));
@@ -1514,7 +1514,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool doctypeNameState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       currentDoctypeToken!.name = currentDoctypeToken!.name?.toAsciiLowerCase();
       state = afterDoctypeNameState;
     } else if (data == '>') {
@@ -1539,7 +1539,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool afterDoctypeNameState() {
     var data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       return true;
     } else if (data == '>') {
       _addToken(currentToken);
@@ -1595,7 +1595,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool afterDoctypePublicKeywordState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       state = beforeDoctypePublicIdentifierState;
     } else if (data == "'" || data == '"') {
       _addToken(ParseErrorToken('unexpected-char-in-doctype'));
@@ -1615,7 +1615,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool beforeDoctypePublicIdentifierState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       return true;
     } else if (data == '"') {
       currentDoctypeToken!.publicId = '';
@@ -1689,7 +1689,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool afterDoctypePublicIdentifierState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       state = betweenDoctypePublicAndSystemIdentifiersState;
     } else if (data == '>') {
       _addToken(currentToken);
@@ -1717,7 +1717,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool betweenDoctypePublicAndSystemIdentifiersState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       return true;
     } else if (data == '>') {
       _addToken(currentToken);
@@ -1743,7 +1743,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool afterDoctypeSystemKeywordState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       state = beforeDoctypeSystemIdentifierState;
     } else if (data == "'" || data == '"') {
       _addToken(ParseErrorToken('unexpected-char-in-doctype'));
@@ -1763,7 +1763,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool beforeDoctypeSystemIdentifierState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       return true;
     } else if (data == '"') {
       currentDoctypeToken!.systemId = '';
@@ -1837,7 +1837,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool afterDoctypeSystemIdentifierState() {
     final data = stream.char();
-    if (isWhitespace(data)) {
+    if (data != null && isWhitespace(data)) {
       return true;
     } else if (data == '>') {
       _addToken(currentToken);
```